inithome/src/rubin/nublado/inithome/provisioner.py

```diff
@@ -0,0 +1,103 @@
+
+    async def provision(self) -> None:
+        """Provision user home directory.
+
+        This is the only public method.  Given an initialized
+        object with UID, GID, and homedir path, if the path does not
+        exist, create it and set it to mode 0700 for the path
+        components created.
+
+        If the path already exists, verify that it is a directory and owned
+        by the right UID and GID; warn if mode is not correct, but don't treat
+        that as a fatal error.
+        """
+        if self.homedir.exists():
+            if not self.homedir.is_dir():
+                raise RuntimeError(
+                    f"{self.homedir} exists but is not a directory"
+                )
+            # Check ownership and permissions
+            stat_results = self.homedir.stat()
+            if stat_results.st_uid != self.uid:
+                raise RuntimeError(
+                    f"{self.homedir} is owned by {stat_results.st_uid}, not "
+                    f"{self.uid}"
+                )
+            if stat_results.st_gid != self.gid:
+                raise RuntimeError(
+                    f"{self.homedir} is owned by group "
+                    f"{stat_results.st_gid}, not {self.gid}"
+                )
+            # We're masking st_mode because BSD and Linux disagree on file
+            # type bit interpretation, and what we care about is rwx------
+            if (stat_results.st_mode & 0o777) != 0o700:
+                logger = logging.getLogger(__name__)
+                logger.warning(
+                    f"{self.homedir} has strange permissions "
+                    f"{oct(stat_results.st_mode & 0o777)} rather than '0o700'"
+                )
+            return
+        # We need to create the directory
+        # Set umask so intermediate paths are created mode 0o700 as well
+        umask(0o077)
+        # Find out how many layers of directory we need to create, for the
+        # chown.
+        first_created = self.homedir
+        while True:
+            if first_created.parent.is_dir() or first_created == Path("/"):
+                break
+            first_created = first_created.parent
+        # Create the homedir (and path leading down to it if needed)
+        self.homedir.mkdir(parents=True)
+        # Climb the path, chown'ing until we get to a layer that previously
+        # existed.  (Neither os.chown() nor shutil.chown() implements "-r")
+        current_dir = self.homedir
+        while True:
+            chown(current_dir, uid=self.uid, gid=self.gid)
+            if current_dir == first_created:
+                break
+            current_dir = current_dir.parent
+
+
+def main() -> None:
+    """Entry point for provisioner.
+
+    Environment variables `NUBLADO_UID` and `NUBLADO_HOME` must be set.
+    if NUBLADO_GID is unset it will get set to NUBLADO_UID.
+    """
+    uid = int(environ["NUBLADO_UID"])  # Fail if unset
+    gid = int(getenv("NUBLADO_GID", uid))  # Default to UID
```

Comment:
GID is now mandatory everywhere in the controller, so this fallback can be removed.